oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/OAuth2TokenWithClaims.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.core;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.springframework.util.Assert;
+
+/**
+ * A composite that contains both an OAuth 2.0 token and any verified claims
+ *
+ * @author Josh Cummings
+ * @since 5.2
+ */
+public class OAuth2TokenWithClaims extends AbstractOAuth2Token implements ClaimAccessor {
```

Comment:
I'm struggling a bit with the addition of this class. I think a lot of this is that the name feels a bit artificial to me. It should be something that is a part of the spec rather than having a specific property on it. I don't think we should have classes that are named based upon properties they contain.

I haven't looked around yet for a more appropriate existing class, but it seems like there should be one. If there isn't one, I believe we should rethink the name to align with something within the spec.

Reply:
I agree with this sentiment. One name I've considered is `OAuth2TokenDetails`. My reasoning is that `UserDetails` contains a user and details about the user, and an `OAuth2TokenDetails` contains a token and details about the token. It also aligns with the fact that both of these would be the result of `getPrincipal`, in their associated contexts.

Reply:
Another option I considered was `IntrospectedToken`. This certainly refers to a spec term, which is good, but it fails at a couple of points:

- First, it's odd that the name indicates how it was validated. `UserDetails`, `Jwt`, and `username` do not indicate how the principal was validated, though the `Authentication` type might, e.g. `UsernamePasswordAuthenticationToken`.
- Second, it pushes this information into web controllers:

```java
@GetMapping
public String request(@AuthenticationPrincipal IntrospectedToken token) {
    // ...
}
```
Which isn't as flexible.

For example, a class like this is handy because the code could now do:

```java
public Jwt extends OAuth2TokenWithClaims
```

And controllers could now do

```java
@GetMapping
public String request(@AuthenticationPrincipal OAuth2TokenWithClaims token)
```

Making the user's code more flexible.

Reply:
My vote is to name it `IntrospectedOAuth2AccessToken` and extend `OAuth2AccessToken`. I think the name also aligns with `IntrospectedOAuth2TokenAuthenticationToken`. However, to make it more aligned we may consider renaming to `IntrospectedOAuth2AccessTokenAuthenticationToken` - bit of a long one though.

Reply:
@jgrandja, I believe `IntrospectedOAuth2AccessToken` is not ideal for the same reasons that `IntropsectedToken` is not ideal.

Reply:
I've updated this class to be named `OAuth2TokenAccessor` since it extends `AbstractOAuth2Token` and implements `ClaimAccessor`. I like this on hierarchical clarity, but still feel that `OAuth2TokenDetails` carries more semantic weight.

I believe there is value in this becoming the parent to `Jwt`, allowing the application to consider the OAuth 2.0 principal without needing to know the manner in which the token was validated.

This is why I'm not keen on `OAuth2AccessTokenDetails` since its parent would be `OAuth2AccessToken`, meaning that I still need to know which validation method was used in order to correctly obtain the principal, e.g. if I'm using `opaque()` then get an `OAuth2AccessTokenDetails`, but if I'm using `jwt()` then get a `Jwt`. This is disappointing since my access patterns at that level will largely be identical.

I like `OAuth2TokenDetails` and `OAuth2TokenAccessor` because they bear out the semantics that this particular token contains unstructured information--it doesn't follow the JWT structure or the JWS structure, though `Jwt`, its subclass, adds the JWT structure to it, like headers and a specialized `ClaimsAccessor`.

I've also considered two other names, which, based on my explanations above, will hopefully clearly be sub-optimal, but I'll include them for completeness:

`StructuredOAuth2Token` - A plain OAuth 2.0 token doesn't have structure, but one that has been introspected at least has the implied structure of attributes and values. Subclasses impose additional structure, in the same way as I already explained.
`AccessibleOAuth2Token` - An adjective version of `OAuth2TokenAccessor`, simply to maintain the concrete class at the end of the name

Reply:
I'm not really feeling `OAuth2TokenAccessor`. Domain objects implicitly have accessors so I feel that it's redundant in the name. As well, the existing classes that implement `ClaimAccessor` do not have `Accessor` in the name either.

I prefer `OAuth2TokenDetails` over `OAuth2TokenAccessor` but I still feel very strongly that we should stick with a name that is referred to in the spec. The obvious is to have `Introspection` or `Introspected` in the name.

After reviewing the spec again, it came to my occurrence that an `IntrospectedOAuth2Token`  should not live outside of an `Authentication` or an Introspection Request/Response flow. I think the following design proposal may address some of your previous concerns @jzheaux as far as naming.

```java
public class IntrospectedOAuth2TokenAuthenticationToken
		extends AbstractOAuth2TokenAuthenticationToken<IntrospectedOAuth2TokenAuthenticationToken.IntrospectedOAuth2Token> {

	public IntrospectedOAuth2TokenAuthenticationToken(IntrospectedOAuth2Token introspectedToken) {
		super(introspectedToken);
	}

	@Override
	public Map<String, Object> getTokenAttributes() {
		return getToken().getAttributes();
	}

	public static class IntrospectedOAuth2Token extends AbstractOAuth2Token {
		private final Map<String, Object> attributes;

		public IntrospectedOAuth2Token(String tokenValue, Instant issuedAt,
										Instant expiresAt, Map<String, Object> attributes) {
			super(tokenValue, issuedAt, expiresAt);
			Assert.notEmpty(attributes, "attributes cannot be empty");
			this.attributes = Collections.unmodifiableMap(new LinkedHashMap<>(attributes));
		}

		public Map<String, Object> getAttributes() {
			return this.attributes;
		}
	}
}
```

Thoughts?

Reply:
I don't like the fact that the name has Token in it twice. This seems to be a problem in more than one place.

Reply:
@rwinch 

> This seems to be a problem in more than one place.

Where is the other place you are referring to?

Reply:
Both `AbstractOAuth2TokenAuthenticationToken` and `IntrospectionOAuth2TokenAuthenticationToken` have token in them twice

Reply:
This is resolved by pushing this down to a discussion about what is the principal in the authentication token, for which I have a ticket.

Reply:
This conversation is resolved with https://github.com/spring-projects/spring-security/issues/6498